flink-ml-lib/src/main/java/org/apache/flink/ml/classification/knn/KnnModel.java

```diff
@@ -0,0 +1,594 @@
+            implements OneInputStreamOperator<Row, Row> {
+
+        public PredictOperator(RichMapFunction<Row, Row> userFunction) {
+            super(userFunction);
+        }
+
+        @Override
+        public void processElement(StreamRecord<Row> streamRecord) throws Exception {
+            Row value = streamRecord.getValue();
+            output.collect(new StreamRecord<>(userFunction.map(value)));
+        }
+    }
+
+    private ResolvedSchema getOutputSchema(
+            ResolvedSchema dataSchema, Map<Param<?>, Object> params, DataType idType) {
+        String[] reservedCols = dataSchema.getColumnNames().toArray(new String[0]);
+        DataType[] reservedTypes = dataSchema.getColumnDataTypes().toArray(new DataType[0]);
+        String[] resultCols = new String[] {(String) params.get(KnnParams.PREDICTION_COL)};
+        DataType[] resultTypes = new DataType[] {idType};
+        return ResolvedSchema.physical(
+                ArrayUtils.addAll(reservedCols, resultCols),
+                ArrayUtils.addAll(reservedTypes, resultTypes));
+    }
+
+    /**
+     * cast data x to t type.
+     *
+     * @param x data.
+     * @param t type.
+     * @return
+     */
+    private static Object castTo(Object x, DataType t) {
+        if (x == null) {
+            return null;
+        } else if (t.equals(DataTypes.BOOLEAN())) {
+            if (x instanceof Boolean) {
+                return x;
+            }
+            return Boolean.valueOf(x.toString());
+        } else if (t.equals(DataTypes.BYTES())) {
+            if (x instanceof Number) {
+                return ((Number) x).byteValue();
+            }
+            return Byte.valueOf(x.toString());
+        } else if (t.equals(DataTypes.INT())) {
+            if (x instanceof Number) {
+                return ((Number) x).intValue();
+            }
+            return Integer.valueOf(x.toString());
+        } else if (t.equals(DataTypes.BIGINT())) {
+            if (x instanceof Number) {
+                return ((Number) x).longValue();
+            }
+            return Long.valueOf(x.toString());
+        } else if (t.equals(DataTypes.FLOAT())) {
+            if (x instanceof Number) {
+                return ((Number) x).floatValue();
+            }
+            return Float.valueOf(x.toString());
+        } else if (t.equals(DataTypes.DOUBLE())) {
+            if (x instanceof Number) {
+                return ((Number) x).doubleValue();
+            }
+            return Double.valueOf(x.toString());
+        } else if (t.equals(DataTypes.STRING())) {
+            if (x instanceof String) {
+                return x;
+            }
+            return x.toString();
+        } else {
+            throw new RuntimeException("unsupported type: " + t.getClass().getName());
+        }
+    }
+
+    /** @return parameters for algorithm. */
+    @Override
+    public Map<Param<?>, Object> getParamMap() {
+        if (null == this.params) {
```

Comment:
This condition might never be true, as all constructors of this class initializes `params`.

Reply:
OK, I will refine it.

Reply:
done